net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
+	return gsf
+}
+
+func (gsf *GraphSyncFetcher) loader(lnk ipldp.Link, lnkCtx ipldp.LinkContext) (io.Reader, error) {
+	asCidLink, ok := lnk.(cidlink.Link)
+	if !ok {
+		return nil, fmt.Errorf("Unsupported Link Type")
+	}
+	block, err := gsf.bs.Get(asCidLink.Cid)
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(block.RawData()), nil
+}
+
+func (gsf *GraphSyncFetcher) storer(lnkCtx ipldp.LinkContext) (io.Writer, ipldp.StoreCommitter, error) {
+	var buffer bytes.Buffer
+	committer := func(lnk ipldp.Link) error {
+		asCidLink, ok := lnk.(cidlink.Link)
+		if !ok {
+			return fmt.Errorf("Unsupported Link Type")
+		}
+		block, err := blocks.NewBlockWithCid(buffer.Bytes(), asCidLink.Cid)
+		if err != nil {
+			return err
+		}
+		return gsf.bs.Put(block)
+	}
+	return &buffer, committer, nil
+}
+
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.GetSingleLayer(ctx, cids, from)
+	if err != nil {
+		return nil, err
+	}
+	startingTipset, err := gsf.LoadTipsetFromCids(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	out := []types.TipSet{startingTipset}
+	isDone, err := done(startingTipset)
+	if err != nil {
+		return nil, err
+	}
+	if !isDone {
+		baseCid := cids[0]
+		recursionDepth := 1
+	makeRecursion:
+		for {
+			cidResponseLayers, err := gsf.GetBlocks(ctx, baseCid, from, recursionDepth)
+			if err != nil {
+				return nil, err
+			}
+			for _, cidResponseLayer := range cidResponseLayers {
+				ts, err := gsf.LoadTipsetFromCids(ctx, cidResponseLayer)
+				if err != nil {
+					return nil, err
+				}
+
+				out = append(out, ts)
+				isDone, err := done(ts)
+				if err != nil {
+					return nil, err
+				}
+
+				if isDone {
+					break makeRecursion
+				}
+				baseCid = cidResponseLayer[0]
+			}
+			if recursionDepth < maxRecursionDepth {
+				recursionDepth *= recursionMultiplier
+			}
+		}
+	}
+	return out, nil
+}
+
+// LoadTipsetFromCids loads and sanitizes an already downloaded TipSet
+// from an array of cids
+func (gsf *GraphSyncFetcher) LoadTipsetFromCids(ctx context.Context, cidResponseLayer []cid.Cid) (types.TipSet, error) {
+	blockResponseLayer := make([]blocks.Block, 0, len(cidResponseLayer))
+	for _, cid := range cidResponseLayer {
+		block, err := gsf.bs.Get(cid)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		blockResponseLayer = append(blockResponseLayer, block)
+	}
+	validatedBlocks, err := sanitizeBlocks(ctx, blockResponseLayer, gsf.validator)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	ts, err := types.NewTipSet(validatedBlocks...)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	return ts, nil
+}
+
+type resultSet struct {
```

Comment:
Please document what this is intended to do. The name seems a bit odd to me - would just "response" work?